enterprise/backend/src/metabase_enterprise/sandbox/models/params/field_values.clj

```diff
@@ -0,0 +1,80 @@
+(ns metabase-enterprise.sandbox.models.params.field-values
+  (:require [clojure.core.memoize :as memoize]
+            [metabase-enterprise.enhancements.ee-strategy-impl :as ee-strategy-impl]
+            [metabase-enterprise.sandbox.api.table :as sandbox.api.table]
+            [metabase.api.common :as api]
+            [metabase.models.field :as field :refer [Field]]
+            [metabase.models.field-values :as field-values :refer [FieldValues]]
+            [metabase.models.params.field-values :as params.field-values]
+            [metabase.public-settings.metastore :as settings.metastore]
+            [metabase.util :as u]
+            [pretty.core :as pretty]
+            [toucan.db :as db]))
+
+(def ^:private ^{:arglist '([user-id last-updated field])} fetch-sandboxed-field-values*
+  (memoize/ttl
+   ;; use a custom key fn for memoization so two Field maps with different keys but the same `:id` still have the same
+   ;; cache key.
+   ^{::memoize/args-fn (fn [[current-user-id updated-at field]]
+                         [current-user-id updated-at (u/the-id field)])}
+   (fn [_ _ field]
+     {:values   (field-values/distinct-values field)
+      :field_id (u/the-id field)})
+   ;; TODO -- shouldn't we return sandboxed human-readable values as well??
+   ;;
+   ;; Expire entires older than 30 days so we don't have entries for users and/or fields that
+   ;; no longer exists hanging around.
+   ;; (`clojure.core.cache/TTLCacheQ` (which `memoize` uses underneath) evicts all stale entries on
+   ;; every cache miss)
+   :ttl/threshold (* 1000 60 60 24 30)))
+
+(defn- fetch-sandboxed-field-values
+  [field]
+  (fetch-sandboxed-field-values*
+   api/*current-user-id*
+   (db/select-one-field :updated_at FieldValues :field_id (u/the-id field))
+   field))
+
+(defn- field-is-sandboxed? [field]
+  (sandbox.api.table/only-segmented-perms? (field/table field)))
```

Comment:
`field/table` does a `(db/select-one 'Table, :id table_id)`. If segmentation of a field is really segmentation of a table, perhaps grouping by table-id and then basing segmentation off of that would save some trips to the db. (instead of the `(group-by (comp boolean field-is-sandboxed?) fields)` below

Reply:
or does the qp store come into play here?

Reply:
Actually, we could just `hydrate` the `:table` for the Fields (Toucan will do a batched hydration here) so we don't end up making a bunch of separate calls